framework/src/meshmodifiers/ElementSubdomainModifierBase.C

```diff
@@ -0,0 +1,662 @@
+    if (std::find(_subdomain_ids_to_reinitialize.begin(),
+                  _subdomain_ids_to_reinitialize.end(),
+                  Moose::ANY_BLOCK_ID) != _subdomain_ids_to_reinitialize.end())
+      _reinitialized_elems.insert(elem_id);
+    else // Reinitialize if new subdomain is in list of subdomains to be reinitialized
+    {
+      const auto & [from, to] = subdomain;
+      if (subdomainIsReinitialized(to) && _old_subdomain_reinitialized)
+        _reinitialized_elems.insert(elem_id);
+      // Only reinitialize if original subdomain is not in list of subdomains
+      else if (subdomainIsReinitialized(to) && !_old_subdomain_reinitialized &&
+               !subdomainIsReinitialized(from))
+        _reinitialized_elems.insert(elem_id);
+      else // New subdomain is not in list of subdomains
+        continue;
+    }
+
+    const auto elem = _mesh.elemPtr(elem_id);
+    for (unsigned int i = 0; i < elem->n_nodes(); ++i)
+      if (nodeIsNewlyReinitialized(elem->node_id(i)))
+        _reinitialized_nodes.insert(elem->node_id(i));
+  }
+}
+
+bool
+ElementSubdomainModifierBase::subdomainIsReinitialized(SubdomainID id) const
+{
+  // Default: any element that changes subdomain is reinitialized
+  if (std::find(_subdomain_ids_to_reinitialize.begin(),
+                _subdomain_ids_to_reinitialize.end(),
+                Moose::ANY_BLOCK_ID) != _subdomain_ids_to_reinitialize.end())
+    return true;
+
+  // Is subdomain in list of subdomains to be reinitialized
+  return std::find(_subdomain_ids_to_reinitialize.begin(),
+                   _subdomain_ids_to_reinitialize.end(),
+                   id) != _subdomain_ids_to_reinitialize.end();
+}
+
+bool
+ElementSubdomainModifierBase::nodeIsNewlyReinitialized(dof_id_type node_id) const
+{
+  // If any of the node neighbors are already reinitialized, then the node is NOT newly
+  // reinitialized.
+  for (auto neighbor_elem_id : _mesh.nodeToElemMap().at(node_id))
+    if (subdomainIsReinitialized(_mesh.elemPtr(neighbor_elem_id)->subdomain_id()))
+      return false;
+  return true;
+}
+
+void
+ElementSubdomainModifierBase::applyIC(bool displaced)
+{
+  _fe_problem.projectInitialConditionOnCustomRange(reinitializedElemRange(displaced),
+                                                   reinitializedBndNodeRange(displaced));
+
+  // Set old and older solutions on the reinitialized dofs to the reinitialized values
+  setOldAndOlderSolutions(_fe_problem.getNonlinearSystemBase(_sys.number()),
+                          reinitializedElemRange(displaced),
+                          reinitializedBndNodeRange(displaced));
+  setOldAndOlderSolutions(_fe_problem.getAuxiliarySystem(),
+                          reinitializedElemRange(displaced),
+                          reinitializedBndNodeRange(displaced));
+
+  // Note: Need method to handle solve failures at timesteps where subdomain changes. The old
+  // solutions are now set to the reinitialized values. Does this impact restoring solutions
```

Comment:
it's a good thought.
We should add an assert here, querying the problem for whether it just failed a time step (if we have such an API)

Reply:
actually, seems like it would be problematic to put an error there. Maybe just a "flagSolutionInvalid" would be enough.
Anyway this is out of scope here